Add extra MCV placement logic

NotoriousPyro opened the pull request:
AI MCVs now place new MCVs near new tiberium fields, away from enemies buildings and its own previous con yards. Similar to a real player.

This modifies the previous logic, to:

1. Get a new random base center that is near a resource field, outside of the centered area of a previous random con yard location. Thus making the AI build outside of previous MCV locations.
2. The normal logic then is used (which selects a random location in the area based on step 1).
3. The MaxBaseRadius around the newly selected location is checked for enemies. If none exist we continue, otherwise we return null.
4. The MaxBaseRadius around the newly selected location is checked for previous construction yards (of the same player) or MCVs. If none exist in that location we continue, otherwise we return null.

Here are a few screenshots of a tweaked AI I created to produce the additional MCVs:
![image](https://user-images.githubusercontent.com/706132/73895994-602c9a80-4879-11ea-8506-8e8f617891ca.png)

![image](https://user-images.githubusercontent.com/706132/73896351-a8988800-487a-11ea-9179-f9144142c191.png)


![image](https://user-images.githubusercontent.com/706132/73896014-6fabe380-4879-11ea-8e52-9922cfe03d69.png)

teinarss committed:
Added widget showing the army for players in spec

teinarss committed:
Updated production spec widget to handle overflow

tovl committed:
Let immovable actors block individual minefield cells.

dragunoff committed:
Calculate position of dropdown arrow glyph based on image width

pchote committed:
Add introduction prompt on first run.

pchote committed:
Change default name from Newbie to Commander.

pchote committed:
Fix subcell crusher logic.

pchote committed:
Set default rally point to an empty path.

pchote committed:
Fix production door not closing if the actor stays on the exit cell.

pchote committed:
Change default player color to dark red.

MustaphaTR committed:
Add PlayLaunchSounds() to SupportPowers which are missing it.

abcdefg30 committed:
Fix a crash in the ReplaceAttackTypeStrafe update rule

everclear72216 committed:
Added display selection option to settings for fullscreen modes.

NotoriousPyro committed:
Add extra MCV placement logic

AI MCVs now place new MCVs near new tiberium fields, away from enemies buildings and its own previous structures. Similar to a real player.

teinarss commented on OpenRA.Mods.Common/Traits/BotModules/McvManagerBotModule.cs:
```diff
@@ -204,10 +218,33 @@ void DeployMcv(IBot bot, Actor mcv, bool move)
 			};
 
-			var baseCenter = GetRandomBaseCenter();
+			CPos baseCenter = GetRandomBaseCenter(distanceToBaseIsImportant);
```
use `var` here.

teinarss commented on OpenRA.Mods.Common/Traits/BotModules/McvManagerBotModule.cs:
```diff
@@ -204,10 +218,33 @@ void DeployMcv(IBot bot, Actor mcv, bool move)
+			baseCenter = bc.Value;
+
+			WPos wPos = world.Map.CenterOfCell(bc.Value);
+			WDist newBaseRadius = new WDist(Info.MaxBaseRadius * 1024);
```
Same here.

teinarss commented on OpenRA.Mods.Common/Traits/BotModules/McvManagerBotModule.cs:
```diff
@@ -51,13 +51,27 @@ public class McvManagerBotModuleInfo : ConditionalTraitInfo
+			var tileset = world.Map.Rules.TileSet;
+			BitArray resourceTypeIndices;
+
+			resourceTypeIndices = new BitArray(tileset.TerrainInfo.Length);
```
You can join these two `var resourceTypeIndices = new BitArray(tileset.TerrainInfo.Length);`

NotoriousPyro commented:
Ah, yes I seem to have missed that as part of the coding standards. I will make these changes later.

NotoriousPyro committed:
Add requested changes from OpenRA team

Change explicit local variables to var

Replace multiple calls to FindActorsInCircle with linq queries

abcdefg30 commented on OpenRA.Mods.Common/Traits/BotModules/McvManagerBotModule.cs:
```diff
@@ -51,13 +51,26 @@ public class McvManagerBotModuleInfo : ConditionalTraitInfo
-		public CPos GetRandomBaseCenter()
+		public CPos GetRandomBaseCenter(bool distanceToBaseIsImportant)
 		{
+			if (distanceToBaseIsImportant == false)
```
This can simplify to `if (!distanceToBaseIsImportant)`.

NotoriousPyro committed:
Add changes recommended by OpenRA team

Replace == false with !condition

Make MCV placement also avoid enemy units to avoid it trying to place nearby enemy squads

NotoriousPyro commented:
I have added that change, and also a change to make the MCV placement avoid any enemy units not only buildings, as the AI was trying to place MCVs next to squads.

abcdefg30 committed:
Fix a crash when a SourceActor does not occupy space

abcdefg30 committed:
Make WarheadArgs.Source nullable

abcdefg30 committed:
Remove a loop closure guard variable

teinarss committed:
Add unique ingame-observer.yaml for D2K

pchote committed:
Extract sysinfo logic to its own class.

pchote committed:
Generalize RallyPointInfo.Offset to support arbitrary length paths.

NotoriousPyro committed:
Update McvManagerBotModule

Disallow the AI from queuing up more than 1 MCV at a time

NotoriousPyro committed:
Add rec. changes by OpenRA team

NotoriousPyro committed:
Merge branch 'mcvplacement' into mcvlimits2

NotoriousPyro committed:
Remove unused setting & update descriptions

Update description for MaxBaseRadius

Remove setting RestrictMCVDeploymentFallbackToBase

Fix AI MCV placement being 'too' random

NotoriousPyro commented:
I think I may have accidentally included a merge from one of my other branches into this PR...

NotoriousPyro committed:
Revert last commit

NotoriousPyro commented:
I'll reopen this commit when I've fixed it up in a separate branch.
